### ISSUE TITLE
Sync dispatch in componentDidMount() does not update component

### ISSUE THREAD

**Issue by roadhump** [gaearon/redux#208]
I've tried v1.0.0-alpha and noticed that this small app https://gist.github.com/roadhump/f40483704e8527f11b07 will not work if I immediately dispatch action result from `componentDidUpdate` but it works if dispatch async (set const `async` to true). It worked both ways in version 0.12.

**Comment by taylorhakes** [comment]
@gaearon I have confirmed this bug. In v1.0.0  the `createConnector` code

```
this.context.store.subscribe(::this.handleChange)
```
was moved from the constructor to the `componentDidMount`. This causes an issue with sync dispatches because the `<Connector>` mounts after the child components and the `this.handleChange` listener is not attached yet. 

To fix this, we will have to move that code back to the constructor or earlier in the component lifecycle, so it will be called before the child components call dispatch.

**Comment by taylorhakes** [comment]
Here is the related issue and pull request. #178 #179

**Comment by johanneslumpe** [comment]
Wasn't this done to prevent setState etc when doing server-side rendering? @gaearon

**Comment by taylorhakes** [comment]
Yes, I linked to the pull request and issue

**Comment by johanneslumpe** [comment]
@taylorhakes What i meant was that moving it back to the constructor or earlier in the lifecycle will possibly have negative implications for server-side rendering.

**Comment by taylorhakes** [comment]
Correct. Another option is to document that componentDidMount dispatches need to be wrapped in `setTimeout` or another async timer

```js
componentDidMount() {
  setTimeout(() => {
    this.props.dispatch(/* ... */);
  }, 0);
}

```

That approach feels very brittle though. Currently React's mounting is synchronous, but it might change in the future.

**Comment by clearjs** [comment]
Why not simply check environment? E.g., with `canUseDOM` from `react/lib/ExecutionEnvironment`. Or am I missing something?

**Comment by johanneslumpe** [comment]
@clearjs is that exported by default? If not, it should be treated as internal file and is off limits. If it's public though, it might work.

**Comment by clearjs** [comment]
It's in the add-ons, so subject to more frequent changes. But the implementation is trivial:

```javascript
var canUseDOM = !!(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
); 
```

Obviously, this would only work in browser, not with react-native, etc.

**Comment by johanneslumpe** [comment]
I think a solution which only works on the browser is not really acceptable though.

**Comment by clearjs** [comment]
Sorry, my last comment was confusing. It does work on Node (returns `false`). Since for react-native server-side rendering isn't an option, this shouldn't be a problem.

**Comment by clearjs** [comment]
Here's a more robust alternative, and would work with react-native as expected: https://www.npmjs.com/package/detect-node

**Comment by johanneslumpe** [comment]
@clearjs ok that detects node. But what about other environments like ruby? You can actually do server-side rendering there too. Or even with php. They will still execute in a JS context, but that will be V8 and not necessarily node I assume. In those cases this would still break I think.

**Comment by clearjs** [comment]
Yep, you're right.

**Comment by taylorhakes** [comment]
Another option. What about forcing the child components to mount after the Connector? Render the children only after componentDidMount on the Connector is called. No need for environment detection.

**Comment by taylorhakes** [comment]
Forgot about server not calling componentDidMount, scratch that idea.

**Comment by gaearon** [comment]
Can you please turn this into a failing test?

**Comment by gaearon** [comment]
Also, can't we do something like

```js
    componentDidMount() {      
      this.unsubscribe = this.context.redux.subscribe(::this.handleChange);

      const nextState = this.selectState(this.props, this.context);
      if (!this.isSliceEqual(this.state.slice, nextState.slice)) {
        this.setState(nextState);
      }
    }
```

**Comment by taylorhakes** [comment]
I threw the example into a failing test. https://github.com/taylorhakes/redux/blob/failed-sync-dispatch/test/fullDispatch.spec.js

**Comment by gaearon** [comment]
Should be fixed in #195 via https://github.com/gaearon/redux/commit/406ca6225b8154afee7d6ea4a389572263bee921.
